kernel/src/transaction.rs

```diff
@@ -0,0 +1,543 @@
+
+        // step three: commit the actions as a json file in the log
+        let json_handler = engine.get_json_handler();
+        match json_handler.write_json_file(&commit_path.location, Box::new(actions), false) {
+            Ok(()) => Ok(CommitResult::Committed(commit_version)),
+            Err(Error::FileAlreadyExists(_)) => Ok(CommitResult::Conflict(self, commit_version)),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Set the operation that this transaction is performing. This string will be persisted in the
+    /// commit and visible to anyone who describes the table history.
+    pub fn operation(&mut self, operation: String) {
+        self.operation = Some(operation);
+    }
+
+    /// WARNING: This is an unstable API and will likely change in the future.
+    ///
+    /// Add commit info to the transaction. This is commit-wide metadata that is written as the
+    /// first action in the commit. The engine data passed here must have exactly one row, and we
+    /// only read one column: `engineCommitInfo` which must be a map<string, string> encoding the
+    /// metadata.
+    ///
+    /// The engine is required to provide commit info before committing the transaction. If the
+    /// engine would like to omit engine-specific commit info, it can do so by passing pass a
+    /// commit_info engine data chunk with one row and one column of type `Map<string, string>`
+    /// that can either be `null` or contain an empty map.
+    ///
+    /// Any other columns in the data chunk are ignored.
+    pub fn commit_info(&mut self, commit_info: Box<dyn EngineData>) {
+        self.commit_info = Some(commit_info.into());
+    }
+}
+
+/// Result after committing a transaction. If 'committed', the version is the new version written
+/// to the log. If 'conflict', the transaction is returned so the caller can resolve the conflict
+/// (along with the version which conflicted).
+// TODO(zach): in order to make the returning of a transcation useful, we need to add APIs to
+// update the transaction to a new version etc.
+#[derive(Debug)]
+pub enum CommitResult {
+    /// The transaction was successfully committed at the version.
+    Committed(Version),
+    /// The transaction conflicted with an existing version (at the version given).
+    Conflict(Transaction, Version),
+}
+
+// given the engine's commit info we want to create commitInfo action to commit (and append more actions to)
+fn generate_commit_info(
+    engine: &dyn Engine,
+    operation: Option<&str>,
+    engine_commit_info: &dyn EngineData,
+) -> DeltaResult<Box<dyn EngineData>> {
+    if engine_commit_info.length() != 1 {
+        return Err(Error::InvalidCommitInfo(format!(
+            "Engine commit info should have exactly one row, found {}",
+            engine_commit_info.length()
+        )));
+    }
+
+    let timestamp: i64 = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
```

Comment:
Timestamp manipulation code is starting to proliferate... and it's very manual/duplicated right now. Maybe it's time to start thinking about an internal API we can use for working with timestamp values?

Reply:
yup +1 I can create a follow-up. And I think it makes testing hard since we have places that actually use local machine time and we need to have deterministic tests

Reply:
#428